test/e2e/nodeportlocal_test.go

```diff
@@ -0,0 +1,589 @@
+		validatePortInRange(t, nplAnnotations, defaultStartPort, defaultEndPort)
+		checkTrafficForNPL(t, data, r, nplAnnotations, clientName)
+
+		data.deletePod(testPodName)
+		checkForNPLRuleInIPTABLES(t, data, r, nplAnnotations, antreaPod, testPodIP.ipv4.String(), false)
+	}
+
+}
+
+// TestNPLChangePortRangeAgentRestart tests NodePortLocal functionalities after changing port range.
+// - Create Multiple Nginx Pods.
+// - Change nplPortRange.
+// - Restart Antrea Agent Pods.
+// - Verify that updated port range is being used for NPL.
+func TestNPLChangePortRangeAgentRestart(t *testing.T) {
+	skipIfNotIPv4Cluster(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	enableNPLInConfigmap(t, data)
+	r := require.New(t)
+
+	k8sUtils, err = NewKubernetesUtils(data)
+	r.Nil(err, "Failed to initialize Kubernetes Utils: %v", err)
+
+	annotation := make(map[string]string)
+	annotation[nplSvcAnnotation] = "true"
+	ipFamily := corev1.IPv4Protocol
+	data.createNginxClusterIPServiceWithAnnotation(false, &ipFamily, annotation)
+
+	node := nodeName(0)
+	var testPods []string
+
+	for i := 0; i < 10; i++ {
+		testPodName := randName("test-pod-")
+		testPods = append(testPods, testPodName)
+		err = data.createNginxPod(testPodName, node)
+		r.Nil(err, "Error Creating test Pod: %v", err)
+	}
+
+	clientName := randName("test-client-")
+	if err := data.createBusyboxPodOnNode(clientName, node); err != nil {
+		t.Fatalf("Error when creating Pod '%s': %v", clientName, err)
+	}
+	_, err = data.podWaitForIPs(defaultTimeout, clientName, testNamespace)
+	if err != nil {
+		t.Fatalf("Error when waiting for IP for Pod '%s': %v", clientName, err)
+	}
+
+	updateNPLPortRangeInConfigmap(t, data, defaultStartPort, defaultEndPort, updatedStartPort, updatedEndPort)
+	defer updateNPLPortRangeInConfigmap(t, data, updatedStartPort, updatedEndPort, defaultStartPort, defaultEndPort)
+
+	antreaPod, _ := data.getAntreaPodOnNode(node)
+
+	for _, testPodName := range testPods {
+		nplAnnotationString, testPodIP := getNPLAnnotation(t, data, r, testPodName)
+		var nplAnnotations []k8s.NPLAnnotation
+		json.Unmarshal([]byte(nplAnnotationString), &nplAnnotations)
+
+		checkForNPLRuleInIPTABLES(t, data, r, nplAnnotations, antreaPod, testPodIP.ipv4.String(), true)
+		validatePortInRange(t, nplAnnotations, updatedStartPort, updatedEndPort)
+		checkTrafficForNPL(t, data, r, nplAnnotations, clientName)
+
+		data.deletePod(testPodName)
+		checkForNPLRuleInIPTABLES(t, data, r, nplAnnotations, antreaPod, testPodIP.ipv4.String(), false)
+	}
+}
+
+// TestNPLPodAddEnableNPL tests Pod creation functionalities for NodePortLocal after Enabling NPL.
+// - Disable NodePortLocal in Antrea.
+// - Create Multiple Nginx Pods.
+// - Enable NodePortLocal in Antrea.
+// - Verify Pod Annotation, IPTABLES rules and traffic to test Pods.
+func TestNPLPodAddEnableNPL(t *testing.T) {
```

Comment:
I am not convinced that the tests where we disable NPL, do stuff, then re-enable NPL again are very useful. Aren't we testing the same code path as during a regular restart? What do you think @tnqn ?

Reply:
We wanted to test a few different types of restarts. 
- The test Pods are added, when NPL is enabled. Then restart Agent. This ensures existing rules are intact after restart.
- Add test Pods when Antrea agent is down (or NPL is disabled), so that we can't program the rules from Pod Add/Updates. After enabling NPL and restarting the Agent, the desired rules should get programmed. This is specially useful for checking the new changes for restartability.

Reply:
You are right @tnqn, as per the current code, we are adding rules asynchronously. But after the PR with changes for  restorability gets in, there would be some some changes https://github.com/vmware-tanzu/antrea/pull/1758. 
We wanted to add comprehensive tests for that.

Reply:
I understand why we have tests with agent restart, these are important to ensure that the new reconciliation logic works. However, IMO a test with the following sequence is not useful: 1) Disable NPL, 2) Create Pod, 3) Enable NPL, 4) Verify. This is the same logic as for this sequence: 1) Enable NPL, 2) Create pod, 3) Verify. This is because nothing special happens when we create Pods when NPL is disabled. So I believe `NPLTestMultiplePods` and `TestNPLPodAddEnableNPL` are essentially duplicates.

Reply:
The logic of reconciliation while booting up in https://github.com/vmware-tanzu/antrea/pull/1758 is different from reconciliation from Pod update events, when Antrea Agent is already running.
If we just restart the Antrea Agent Pod when the rules have been already programmed, after boot up, we would only verify if the rules are correct or not, not add/delete the rules. To make sure - that part of the code kicks in - we have to ensure that the rules are not present while we are booting up.

Reply:
That doesn't really clarify it for me, but maybe I am misunderstanding the scope of #1758. IMO, #1758 should only deal with Pods with an existing NPL Annotation and make sure that the iptables state matches these exiting annotations (by adding / deleting / modifying iptables rules as needed). On restart, Pods without an annotation should be handled by the normal event handlers, after that reconciliation process. From that point of view, as Quan phrased it, there should be  no "difference on how NPL agent handles the Pods created before NPL feature is enabled and the ones created after it". But #1758 hasn't been updated since the original round of reviews, so I will get back to this PR after #1758 is finalized.

Reply:
Yeah @antoninbas I get it now. We can't handle Pods without annotation as part of the code for handling restart. I'll remove these tests.